crates/bevy_ecs/src/term_query/builder.rs

```diff
@@ -0,0 +1,84 @@
+use std::marker::PhantomData;
+
+use crate::{
+    component::ComponentId,
+    prelude::{Component, With, Without, World},
+};
+
+use super::{ComponentTerm, QueryTermGroup, Term, TermQueryState, TermVec};
+
+pub struct QueryBuilder<'w, Q: QueryTermGroup = ()> {
+    terms: TermVec<Term>,
+    current_term: usize,
+    world: &'w mut World,
+    _marker: PhantomData<Q>,
+}
+
+impl<'w, Q: QueryTermGroup> QueryBuilder<'w, Q> {
+    pub fn new(world: &'w mut World) -> Self {
+        let mut terms = TermVec::new();
+        Q::init_terms(world, &mut terms);
+        Self {
+            current_term: terms.len(),
+            terms,
+            world,
+            _marker: PhantomData::default(),
+        }
+    }
+
+    pub fn term<T: QueryTermGroup>(&mut self) -> &mut Self {
+        T::init_terms(self.world, &mut self.terms);
+        self
+    }
+
+    pub unsafe fn term_at(&mut self, index: usize) -> &mut Self {
+        self.current_term = index;
+        self
+    }
+
+    pub fn with<T: Component>(&mut self) -> &mut Self {
+        self.term::<With<T>>()
+    }
+
+    pub fn with_id(&mut self, id: ComponentId) -> &mut Self {
+        self.terms.push(Term::Component(ComponentTerm::with(id)));
+        self
+    }
+
+    pub fn without<T: Component>(&mut self) -> &mut Self {
+        self.term::<Without<T>>()
+    }
+
+    pub fn without_id(&mut self, id: ComponentId) -> &mut Self {
+        self.terms.push(Term::Component(ComponentTerm::without(id)));
+        self
+    }
+
+    pub fn ref_id(&mut self, id: ComponentId) -> &mut Self {
+        self.terms.push(Term::Component(ComponentTerm::read_id(id)));
+        self
+    }
+
+    pub fn mut_id(&mut self, id: ComponentId) -> &mut Self {
+        self.terms
+            .push(Term::Component(ComponentTerm::write_id(id)));
+        self
+    }
+
+    pub fn set<T: Component>(&mut self) -> &mut Self {
```

Comment:
The naming here isn't clear. Even with the code in front of me I have a hard time following what exactly this method does.

Reply:
That's very fair, `set` and `set_id` are intended to be used with dynamic pointer fetches like `Ptr` where you don't know the type at compile time.

Especially when combined with the other function names that map directly to terms this one is very misleading. I'm struggling to come up with succinct descriptive names, perhaps `set_ptr_component`, `set_dynamic` or `set_dynamic_component`.

Reply:
This is adding a dynamic component access right? Perhaps `add_dynamic_access` or something?